### ISSUE TITLE
getServerSideProps TypeError: Cannot convert undefined or null to object at Function.keys (<anonymous>) at renderToHTML

### ISSUE THREAD

**Issue by markjackson02** [vercel/next.js#15525]
# Bug report

## Describe the bug

When I switch from using `getInitialProps` to `getServerSideProps`, I get this error in the console:

```
TypeError: Cannot convert undefined or null to object
    at Function.keys (<anonymous>)
    at renderToHTML (C:\Users\markj\workspace\project\node_modules\next\dist\next-server\server\render.js:39:78)
    at async C:\Users\markj\workspace\project\node_modules\next\dist\next-server\server\next-server.js:68:285
    at async C:\Users\markj\workspace\project\node_modules\next\dist\next-server\server\next-server.js:67:142
    at async DevServer.renderToHTMLWithComponents (C:\Users\markj\workspace\project\node_modules\next\dist\next-server\server\next-server.js:91:254)
    at async DevServer.renderToHTML (C:\Users\markj\workspace\project\node_modules\next\dist\next-server\server\next-server.js:92:254)
    at async DevServer.renderToHTML (C:\Users\markj\workspace\project\node_modules\next\dist\server\next-dev-server.js:22:539)
    at async DevServer.render (C:\Users\markj\workspace\project\node_modules\next\dist\next-server\server\next-server.js:48:236)
    at async Object.fn (C:\Users\markj\workspace\project\node_modules\next\dist\next-server\server\next-server.js:35:852)
    at async Router.execute (C:\Users\markj\workspace\project\node_modules\next\dist\next-server\server\router.js:28:28)
    at async DevServer.run (C:\Users\markj\workspace\project\node_modules\next\dist\next-server\server\next-server.js:44:494)
    at async DevServer.handleRequest (C:\Users\markj\workspace\project\node_modules\next\dist\next-server\server\next-server.js:13:133)
```

Does not generate error: 
```
myPage.getInitialProps = async ({ res }) => {
        res.statusCode = 200
        res.setHeader('Content-Type', 'application/json')
        res.end()
};
```

Generates error:
```
export async function getServerSideProps ({ res }) {
        res.statusCode = 200
        res.setHeader('Content-Type', 'application/json')
        res.end() 
};
```

Note that the error is generated AFTER the response is sent to the client.

## To Reproduce

https://codesandbox.io/s/next-js-getserversideprops-typeerror-example-xz8jo

Load index page and check console

## Expected behavior

Rendering process should not continue once res.end() is called.

## System information

- Version of Next.js: 9.4.4
- Version of Node.js: 12.16.3 locally + not sure what version CodeSandbox uses.

**Comment by markjackson02** [comment]
I was able to figure out my issue based on the [API Routes with REST](https://github.com/vercel/next.js/tree/canary/examples/api-routes-rest) example.

1) `pages/api` folder seems to be a special routes folder that is handled differently (it does not expect files to output a component). I moved my file into this folder and that partly resolved the issue.

2) Rather than taking a `context` object, the default export in an API route takes `(req, res)` as separate arguments instead of a `({ req, res})` context object.
